### ISSUE TITLE
Custom Context classes lost in field resolution

### ISSUE THREAD

**Issue by Mange** [rmosolgo/graphql-ruby#2060]
According to the documentation, a schema should be able to set `context_class` to get a custom class for contexts in a query.

This seems to work, but only at the root level. As soon as a field is resolved, the context turns into a `GraphQL::Query::Context::FieldResolutionContext` which is missing the extra methods and is not delegating to the parent context.

How is this supposed to work?

```ruby
require "graphql"

class CustomContext < GraphQL::Query::Context
  def custom_method
    self[:test_value] || "default"
  end
end

QueryRoot = GraphQL::ObjectType.define do
  name "Query"
  field :simple, types.String, resolve: ->(_, _, ctx) { ctx.custom_method }
end

class Schema < GraphQL::Schema
  query QueryRoot
  context_class CustomContext
end

query = GraphQL::Query.new(
  Schema,
  "query { simple }",
  context: {test_value: "hello world"},
)

p query.result
```

<details>
<summary>Stack trace</summary>

```
test.rb:11:in `block (2 levels) in <top (required)>': undefined method `custom_method' for #<GraphQL Context @ Query.simple> (NoMethodError)
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/field.rb:248:in `resolve'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/execute.rb:295:in `call'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/schema/middleware_chain.rb:49:in `invoke_core'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/schema/middleware_chain.rb:38:in `invoke'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/execute.rb:108:in `block in resolve_field'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/tracing.rb:62:in `block in trace'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/tracing.rb:76:in `call_tracers'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/tracing.rb:62:in `trace'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/execute.rb:107:in `resolve_field'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/execute.rb:71:in `block in resolve_selection'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/execute.rb:64:in `each'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/execute.rb:64:in `resolve_selection'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/execute.rb:36:in `block in resolve_root_selection'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/tracing.rb:62:in `block in trace'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/tracing.rb:76:in `call_tracers'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/tracing.rb:62:in `trace'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/execute.rb:32:in `resolve_root_selection'        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/multiplex.rb:108:in `begin_query'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/multiplex.rb:81:in `block in run_as_multiplex'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/multiplex.rb:80:in `map'        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/multiplex.rb:80:in `run_as_multiplex'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/multiplex.rb:69:in `block (2 levels) in run_queries'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/multiplex.rb:174:in `block in instrument_and_analyze'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/instrumentation.rb:29:in `block (2 levels) in apply_instrumenters'        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/instrumentation.rb:46:in `block (2 levels) in each_query_call_hooks'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/instrumentation.rb:41:in `each_query_call_hooks'        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/instrumentation.rb:45:in `block in each_query_call_hooks'        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/instrumentation.rb:70:in `call_hooks'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/instrumentation.rb:44:in `each_query_call_hooks'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/instrumentation.rb:27:in `block in apply_instrumenters'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/instrumentation.rb:70:in `call_hooks'        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/instrumentation.rb:26:in `apply_instrumenters'        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/multiplex.rb:166:in `instrument_and_analyze'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/multiplex.rb:68:in `block in run_queries'        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/tracing.rb:62:in `block in trace'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/tracing.rb:76:in `call_tracers'        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/tracing.rb:62:in `trace'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/execution/multiplex.rb:58:in `run_queries'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/query.rb:168:in `block in result'        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/query.rb:357:in `with_prepared_ast'
        from /home/mange/.rvm/gems/ruby-2.4.4@hemnet-ng/gems/graphql-1.8.8/lib/graphql/query.rb:167:in `result'        from test.rb:25:in `<top (required)>'
```

</details>

Changing the resolver to use `ctx[:test_value]` works, as do `ctx.parent.custom_method`.

Am I supposed to always call `parent` on it? In that case: Do I have to call `#parent` recursively in some cases?

**Comment by rmosolgo** [comment]
Sorry, this is a bit of odd forwards/backwards compatibility issues. 

TL;DR: custom context classes are supported for method-style resolvers, but not `->(obj, args, ctx)`-style resolvers. 

------

Whenever a resolver in the form of `-> (obj, args, ctx)` is used, a special, field-level `ctx` object is created _just for that resolver_. 

For large queries, this doesn't scale well, where lots of objects are created for no real purpose. 

So, the new class-based API made two changes: 

- Use methods instead of `-> (obj, args, ctx)` for resolves; and 
- Don't create new `ctx` objects for each resolver, instead, always reuse the root context 

Building on that, since all resolvers were using the same context object, it was trivial to introduce custom context classes. 

It's harder for the legacy-style `->(obj, args, ctx)` resolvers because FieldResolutionContext is a subclass of the root-level context.

**Comment by Mange** [comment]
Thank you for your reply. We'll try to migrate away from the old style to the new one.

For now I'm monkey patching the new class to delegate all my custom methods too.

```ruby
class CustomContext
  def foo() … end
  def bar() … end
end

# Monkey patch to work around issue described in
# https://github.com/rmosolgo/graphql-ruby/issues/2060
class GraphQL::Query::Context::FieldResolutionContext
  def_delegators :@context,
    :foo,
    :bar
end
```

With a viable workaround and a long-term fix available, I'm closing this.

**Comment by Papipo** [comment]
How can I solve this in the case of field instrumentation? I am using it to allow/disallow access to some fields.

```rb
module AccessInstrumentation
  def self.instrument(_type, field)
    field.redefine do
      resolve ->(obj, args, ctx) {
        ctx.current_ability.can?(:read, obj, field.name) ? field.resolve_proc : nil
      }
    end
  end
end
```

I get `undefined method `current_ability' for #<GraphQL Context @ Query.item>` <- The Context class is not my custom class but the default one.

If there is any other way to hook into field resolution apart from field instrumentation, please let me know. Thanks!
